### ISSUE TITLE
Move build-tool dependency logic earlier, into package promotion / snapshot loading 

### ISSUE THREAD

**Issue by mgsloan** [commercialhaskell/stack#3777]
See this WIP branch - https://github.com/commercialhaskell/stack/tree/fix-finding-promoted-build-tools-3733 - it is intended to solve #3733, but it doesn't quite work.

What it does is add info about executables to every member of `SourceMap`, such that the tool map can be computed from the source map.

However, this doesn't quite cut it, with this change, doing `stack build hlint` in stack's directory triggers this assertion in ConstructPlan:

```haskell
                        -- An assertion to check for a recurrence of
                        -- https://github.com/commercialhaskell/stack/issues/345
                        (assert (destLoc == piiLocation ps) destLoc)
```

The issue is that `happy` is treated as a dep of `haskell-src-exts`, but the snapshot promotion logic thinks it can be installed to the snapshot DB.

I think the solution is to move the resolution of build tool ExeName to PackageName into snapshot loading (possibly in `calculatePackagePromotion`?), deferring any warnings associated with resolving these.  These warnings can later be outputted if used in plan construction.  This would allow for build tool dependencies to cause package promotion.

What do you think, @snoyberg ?

**Comment by mgsloan** [comment]
Note: This should also resolve the following issues https://github.com/commercialhaskell/stack/issues/3195

Another thought I had is, well, maybe we should entirely do away with package promotion by doing implicit snapshots?  Might be better to just go ahead and do implicit snapshots while keeping this issue in mind.

**Comment by snoyberg** [comment]
You're still in about implicit snapshots 😀. They won't actually work here. Local packages really must be local, not part of a snapshot. We need the promotion logic.

Other than that, the plan sounds good.

**Comment by mgsloan** [comment]
Good point, that is true with how we've discussed implicit snapshots in the past.  However, I'm not sure it's necessary to separate snapshot / local DBs https://github.com/commercialhaskell/stack/issues/3783
